### PR TITLE
Affine tasks

### PR DESCRIPTION
**Pull request by Swoorup**
Fixes https://github.com/demystifyfp/FsToolkit.ErrorHandling/issues/106

along with few other fixes

* Use affine tasks 
* Remove side by side github workflow hack

### CODE REVIEW

**Comment by Swoorup** [comment]
@NinoFloris any reasons why methods like `task.Return` from ply return `Ply` type instead of tasks?

**Comment by TheAngryByrd** [comment]
Ugh that side-by-side hack isn't working anymore 😢

**Comment by TheAngryByrd** [comment]
> Ugh that side-by-side hack isn't working anymore 😢

Oh nice it was actually fixed and here's an example https://github.com/actions/setup-dotnet#usage

**Comment by NinoFloris** [comment]
@Swoorup this is just the 'internal' exchange type, same goes for Taskbuilder (which has Step), it's the lowest common denominator between ValueTask and Task.

**Comment by Swoorup** [comment]
Should fstoolkit be using ply as intermediate type at all, or always use task? I am finding it hard to work out compiler messages. The test are failing to build but the library builds.

**Comment by Swoorup** [comment]
It appears fake doesn't allow changing the target to net50. Other than that this is ready for review. @TheAngryByrd

**Comment by Swoorup** [comment]
Not sure how to fix the tooling issue
FSharp.Compiler.CompileOps+FileNameNotResolved' was thrown

**Comment by Swoorup** [comment]
Reverted to just fix the linked issue

**Review by TheAngryByrd — APPROVED** [review]
Looks good to me. @NinoFloris wanna give it a once over as well?

File: src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs [hunk 1]
```diff
@@ -57,63 +56,66 @@ module TaskOptionCE =
...
         member inline _.Combine
           (computation1: Ply<Option<'T>>,
            computation2: unit -> Ply<Option<'U>>)
-          : Ply<Option<'U>> = uply {
-            match! computation1 with
-            | None -> return None
-            | Some _ -> return! computation2()
-          }
+          : Ply<Option<'U>> =
+            task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:65]: revert

File: src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs [hunk 2]
```diff
@@ -57,63 +56,66 @@ module TaskOptionCE =
...
-                match! computation() with
-                | Some _ as o ->
-                  result <- o
-                | None ->
-                  result <- None
-                  fin <- true
-              return result
-            }
+            : Ply<Option<'U>> =
+              task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:92]: revert

File: src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs [hunk 3]
```diff
@@ -57,63 +56,66 @@ module TaskOptionCE =
...
-                match! binder enumerator.Current with
-                | Some _ as o ->
-                  result <- o
-                | None ->
-                  result <- None
-                  fin <- true
-              return result
-            }
+            : Ply<Option<'U>> =
+              task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:107]: revert

File: src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs [hunk 4]
```diff
@@ -30,66 +29,69 @@ module TaskResultCE =
...
     member inline _.Combine
         (computation1: Ply<Result<unit, 'TError>>,
          computation2: unit -> Ply<Result<'U, 'TError>>)
-        : Ply<Result<'U, 'TError>> = uply {
-          match! computation1 with
-          | Error e -> return Error e
-          | Ok _ -> return! computation2()
-        }
+        : Ply<Result<'U, 'TError>> =
+          task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs:43]: revert

File: src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs [hunk 5]
```diff
@@ -30,66 +29,69 @@ module TaskResultCE =
...
-          while not fin && guard() do
-            match! computation() with
-            | Ok x -> x
-            | Error _ as e ->
-              result <- e
-              fin <- true
-          return result
-        }
+        : Ply<Result<unit, 'TError>> =
+          task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs:70]: revert

File: src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs [hunk 6]
```diff
@@ -30,66 +29,69 @@ module TaskResultCE =
...
-          while not fin && enumerator.MoveNext() do
-            match! binder enumerator.Current with
-            | Ok x -> x
-            | Error _ as e ->
-              result <- e
-              fin <- true
-          return result
-        }
+        : Ply<Result<unit, 'TError>> =
+          task {
```
Comment (NinoFloris) [inline src/FsToolkit.ErrorHandling.TaskResult/TaskResultCE.fs:84]: revert

**Review by NinoFloris — CHANGES_REQUESTED** [review]
The switch looks good. 

I'm not quite sure about all the renames from uply to task, as the return type that you get is not `Task` but `Ply`, it doesn't really aid readability this way but I'm not against it either.

One important change that should happen is every place where `uply { ... }` was swapped for `task { ... } |> task.ReturnFrom` should get reverted, this adds needless speedbump allocations going to Task + methodbuilders etc to then go straight back to Ply. 

The idea of uply {} is for it to be used as an internal type of the CE methods as much as possible, only to be mapped to the final type Task/ValueTask/Async in `Run`.

Either way, thanks for doing this :)

@TheAngryByrd I'm all for this change but there was a reason we didn't do this during the conversion to Ply in the first place. It should be very clear to users of the library currently depending on these CEs to automatically apply `ConfigureAwait(false)` that they can no longer expect it to do so. Meaning at every bind where they previously expected the CE to ignore any sync context or task scheduler it's now up to users to sprinkle `ConfigureAwait(false)` on those tasks before binding. If not then this release is bound to create deadlocks for users of gui or older aspnet frameworks.

**Comment by TheAngryByrd** [comment]
> @TheAngryByrd I'm all for this change but there was a reason we didn't do this during the conversion to Ply in the first place. It should be very clear to users of the library currently depending on these CEs to automatically apply ConfigureAwait(false) that they can no longer expect it to do so. Meaning at every bind where they previously expected the CE to ignore any sync context or task scheduler it's now up to users to sprinkle ConfigureAwait(false) on those tasks before binding. If not then this release is bound to create deadlocks for users of gui or older aspnet frameworks. 

Yeah I'm sure I'm gonna get an issue about this problem too. Might have to make a separate package with unaffine vs affine but we'll see if people complain first before I go fixing problems that may not exist.  Additionally the older versions aren't going anywhere for now.

**Review by TheAngryByrd — CHANGES_REQUESTED** [review]
Please address Nino's comments.

File: src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs (hunk 1 above)
Comment (Swoorup) [reply src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:65]: @NinoFloris I was under the impression the unsafe version ignores the context as well? Or are they just restricted to all safe or unsafe coming from non-affine modules?
https://github.com/crowded/ply#execution-bubble

Comment (NinoFloris) [reply src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:65]: The execution context is not the same as the sync context/scheduler. The unsafe CEs don't capture the previous execution context at the start of the CE (captured to be put back when the CE sets a result on the Task), which can cause AsyncLocals to leak into the outer scope, that's the gist of safe vs unsafe. However when Run uses a safe builder to execute the internal ply returning continuation the execution context will be properly captured and reset by that safe builder.

For the sync context/scheduler all that matters is Affine/NonAffine.

Comment (Swoorup) [reply src/FsToolkit.ErrorHandling.TaskResult/TaskOptionCE.fs:65]: Thanks for clearing that up

**Comment by Swoorup** [comment]
@NinoFloris @TheAngryByrd Have addressed the feedback

**Comment by TheAngryByrd** [comment]
Thank you!

**Review by TheAngryByrd — APPROVED** [review]

**Comment by TheAngryByrd** [comment]
@Swoorup Take `2.0.0-beta002` for a spin

**Comment by Swoorup** [comment]
Pre 2.0.0-beta002, the following code would give 

```fsharp
#r "nuget: Ply"
#r "nuget: FsToolkit.ErrorHandling.TaskResult,2.0.0-beta002"

open System
open System.Threading
open System.Threading.Tasks
open FSharp.Control.Tasks
open FsToolkit.ErrorHandling
open FsToolkit.ErrorHandling

let a () =
  task {
    printfn "From taskCE Scheduler Id: %A" TaskScheduler.Current.Id
    let! a = 
      taskResult { 
        do! Task.Delay(TimeSpan.FromSeconds(1.0)) 
        printfn "From taskResultCE Scheduler Id: %A" TaskScheduler.Current.Id
      }
    return ()
  } 

let Scheduler = ConcurrentExclusiveSchedulerPair(TaskScheduler.Default, 1, 1)
printfn "Main Scheduler ID: %A" TaskScheduler.Current.Id
printfn "Scheduler ID: %A" Scheduler.ExclusiveScheduler.Id

Task.Factory.StartNew(a, CancellationToken.None, TaskCreationOptions.None, Scheduler.ExclusiveScheduler) |> ignore
Console.ReadLine()
```

#### Pre 2.0.0-beta002
image

#### 2.0.0-beta002
image

Works well now.

**Pull request merged**